internal/cmd/generateinterceptor/main.go

```diff
@@ -0,0 +1,412 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Code generated by go generate; DO NOT EDIT.
+
+package converter
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	
+	workflowservicepb "go.temporal.io/api/workflowservice/v1"
+	commonpb "go.temporal.io/api/common/v1"
+	historypb "go.temporal.io/api/history/v1"
+	commandpb "go.temporal.io/api/command/v1"
+	failurepb "go.temporal.io/api/failure/v1"
+	workflowpb "go.temporal.io/api/workflow/v1"
+	querypb "go.temporal.io/api/query/v1"
+)
+
+type serviceInterceptor struct {
+	encoders []PayloadEncoder
+}
+
+func (s *serviceInterceptor) encodePayload(payload *commonpb.Payload) error {
+	for i := len(s.encoders) - 1; i >= 0; i-- {
+		if err := s.encoders[i].Encode(payload); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *serviceInterceptor) encodePayloads(payloads *commonpb.Payloads) error {
+	for _, payload := range payloads.GetPayloads() {
+		if err := s.encodePayload(payload); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) decodePayload(payload *commonpb.Payload) error {
+	for _, encoder := range s.encoders {
+		if err := encoder.Decode(payload); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *serviceInterceptor) decodePayloads(payloads *commonpb.Payloads) error {
+	for _, payload := range payloads.GetPayloads() {
+		if err := s.decodePayload(payload); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) process(encode bool, objs ...interface{}) error {
+	for _, obj := range objs {
+		switch o := obj.(type) {
+			case *commonpb.Payload:
+				if encode {
+					if err := s.encodePayload(o); err != nil { return err }
+				} else {
+					if err := s.decodePayload(o); err != nil { return err }
+				}
+			case *commonpb.Payloads:
+				if encode {
+					if err := s.encodePayloads(o); err != nil { return err }
+				} else {
+					if err := s.decodePayloads(o); err != nil { return err }
+				}
+			case map[string]*commonpb.Payload:
+				for _, x := range o { if err := s.process(encode, x); err != nil { return err } }
+{{range $type, $record := .}}
+		{{if $record.Slice}}
+			case []{{$type}}:
+				for _, x := range o { if err := s.process(encode, x); err != nil { return err } }
+		{{end}}
+		{{if $record.Map}}
+			case map[string]{{$type}}:
+				for _, x := range o { if err := s.process(encode, x); err != nil { return err } }
+		{{end}}
+			case {{$type}}:
+				if o == nil { continue }
+				if err := s.process(
+					encode,
+					{{range $record.Methods -}}
+						o.{{.}}(),
+					{{end}}
+				); err != nil { return err }
+{{end}}
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) encode(objs ...interface{}) error {
```

Comment:
Don't really need these if you're willing to pass the bool from inside the interceptor